frontend/appflowy_tauri/src/appflowy_app/components/database/Database.tsx

```diff
@@ -0,0 +1,50 @@
+import { useRef } from 'react';
+import { proxy, useSnapshot } from 'valtio';
+import { DatabaseLayoutPB } from '@/services/backend';
+import { DndContext, DndContextDescriptor, useAutoScrollOnEdge } from './_shared';
+import { VerticalScrollElementRefContext, DatabaseContext, DatabaseUIStateDescriptor, DatabaseUIState } from './database.context';
+import { useViewId, useConnectDatabase, } from './database.hooks';
+import { Grid } from './grid';
+
+export const Database = () => {
+  const viewId = useViewId();
+  const verticalScrollElementRef = useRef<HTMLDivElement>(null);
+  const database = useConnectDatabase(viewId);
+  const snapshot = useSnapshot(database);
+  const dndContext = useRef(proxy<DndContextDescriptor>({
+    dragging: undefined,
+  }));
+  const uiState = useRef<DatabaseUIStateDescriptor>(proxy({
```

Comment:
I suggest encapsulating selector and dispacth, and not allowing any fields to be modified anywhere, which can easily lead to problems.

Reply:
function useSelectorUIState() {
    const uiState = useContext(DatabaseUIState);
    return useSnapshot(uiState);
}

function useUpdate(filedName: string, value: string) {
     const uiState = useContext(DatabaseUIState);
     uiState[filedName] = value;
}

function useDispatchUIState() {
     const uiState = useContext(DatabaseUIState);
     const dispatch = (state) => {
         Object.keys(uiState).forEach(key => uiState[key] = state[key])
     }
}

function createUIState() {
   return proxy({
       enableVerticalAutoScroll: false,
       enableHorizontalAutoScroll: false,
   });
}